File: j2k/src/org/jetbrains/jet/j2k/ast/ClassType.java
package org.jetbrains.jet.j2k.ast;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.jet.j2k.util.AstUtil;

import java.util.Collections;
import java.util.List;

/**
 * @author ignatov
 */
public class ClassType extends Type {
  private final Identifier myType;
  private final List<? extends Element> myParameters;

  public ClassType(Identifier type, List<? extends Element> parameters, boolean nullable) {
    myType = type;
    myParameters = parameters;
    myNullable = nullable;
  }

  public ClassType(Identifier type, List<? extends Element> parameters) {
    myType = type;
    myParameters = parameters;
  }

  public ClassType(Identifier type) {
    myType = type;
    myNullable = false;
    myParameters = Collections.emptyList();
  }

  @NotNull
  @Override
  public String toKotlin() {
    String params = myParameters.size() == 0 ? EMPTY : "<" + AstUtil.joinNodes(myParameters, COMMA_WITH_SPACE) + ">";
    return myType.toKotlin() + params + isNullableStr();
  }
}

File: src/org/jetbrains/jet/j2k/ast/types/ClassType.kt
package org.jetbrains.jet.j2k.ast.types

import org.jetbrains.jet.j2k.ast.Element
import org.jetbrains.jet.j2k.ast.Identifier
import java.util.Collections
import java.util.ArrayList

public open class ClassType(val `type` : Identifier, val parameters : List<Element>, nullable : Boolean) : Type(nullable) {
    public override fun toKotlin() : String {
        // TODO change to map() when KT-2051 is fixed
        val parametersToKotlin = ArrayList<String>()
        for(val param in parameters) {
            parametersToKotlin.add(param.toKotlin())
        }
        var params : String = if (parametersToKotlin.size() == 0)
            ""
        else
            "<" + parametersToKotlin.makeString(", ") + ">"
        return `type`.toKotlin() + params + isNullableStr()
    }


    public override fun convertedToNotNull() : Type = ClassType(`type`, parameters, false)
}
